### ISSUE TITLE
Cancel server-side queries

### ISSUE THREAD

**Issue by ChristianGruen** [BaseXdb/basex#1211]
Inspired by #989 and @apb2006's async module: 

It would be nice if running XQuery expressions could be canceled by another thread. See the following example: If a client is sending search requests for every typed in character, usually only the last search results are relevant. It should then be possible to cancel an obsolete query. This could possibly be realized with forked queries:

* `/restxq/start-query?term=abc&last-id=...`:
cancels a query thread with the specified id, starts a new (forked) query, returns id of new query thread
* `/restxq/get-result?query-id=...`: returns result of forked query, or raises an error query was interrupted

Three new XQuery functions could be sufficient (they may as well be moved to another module, e.g. `http://basex.org/modules/xquery/async`):

* `xquery:fork($query as xs:string, $opts as map(*)) as xs:string`: Starts a new query in another thread and returns a query id
* `xquery:cancel($id as xs:string)`: Cancels a query. Raises an error if query could not be canceled.
* `xquery:get($id as xs:string) as item()*`: Returns the result of the query with the specified id. Raises an error if the query was canceled or does not exist.

The RESTXQ code would then look as follows:

```xquery
declare function local:start-query($term, $last-id) as xs:string {
  xquery:cancel($last-id),
  xquery:fork(...xquery expression for retrieving results...)  
};

declare function local:get-result($query-id) as item()* {
  xquery:get($query-id)
};
```

**Comment by james-jw** [comment]
This would be great! Up vote!

It would also be nice to have a way to pass in a `id` of your choosing, so you can easily cache comparable queries from disjunct clients. In the above examples only the single client can leverage this highly powerful knowledge.

For example:
```xquery
xquery:cancel('term=abc&last-id=34')
xquery:fork('... expression ...', 'term=abc&last-id=34')
```

Also, and maybe this is overkill, but being able to leverage the promise pattern and build chains of XQuery transformation logic with the: `then, done, fail, always, when` pipeline model would be amazing!

**Comment by apb2006** [comment]
I also believe in leveraging the promise pattern :-)

In @ChristianGruen's example `local:get-result($query-id)` implies that BaseX has cached the result somewhere (and is therefore responsible for providing some management strategy?)   Using the `then` pattern means this is not required. This code shows the idea
https://github.com/Quodatum/async/blob/master/src/samples/filelist.xq#L5 The result is available in execution of the `then` (here currently name 'fulfilled' ) as an external variable. This variable is named `$value` currently. These names will change and I doubt it actually works at all at the moment.

**Comment by ChristianGruen** [comment]
Hi! I’ve added an experimental (but very simple) function to our Session Module, which will enable you to interrupt server-side queries in the same HTTP session and with the same user-defined key. This is handy if you know that will never need the result of a launched request, as it may e.g. have got obsolete because of a more recent user request:

Here is a little example:

```xquery
import module namespace session = "http://basex.org/modules/session"; 

declare %rest:path('/query') function local:register() {
  session:register-query('EXPENSIVE'),
  (: the expensive query :)
  (1 to 1000000000000)[. = 0]
};

(:~ Optional service to stop the expensive query. :)
declare %rest:path('/stop') function local:stop() {
  session:register-query('EXPENSIVE')
};
```

By `/query` in the browser, you will launch an expensive query. If the same URL is called in a second browser tab, the first query process will be canceled and „replaced“ with the new query (and `400 Interrupted` will be returned as response for the first request). By calling `.../restxq/stop` in a third tab, the second process will be canceled as well, and an empty sequence will immediately be returned.

Obviously this solution is very basic, and not as powerful as the idea proposed in the initial comment, but I believe it allows us to speed up a fairly common search pattern: If each key click in a client form triggers a new search, we are usually only interested in the most recent search result. We could get rid of client-side search delays, and those kind of checks which are used to find out what is the most recent search results. Looking forward to your feedback!

**Comment by james-jw** [comment]
Awesome addition!

**Comment by ChristianGruen** [comment]
@james-jw: Thanks!

I’m wondering how to properly name the function (I guess that `register query` isn’t that self-describing). Alternatives could be `watch-query`, `monitor-query`… Suggestions are welcome!

**Comment by micheee** [comment]
Cool feature :+1: 
Yet another naming suggestion: `rest:supersede`

**Comment by ChristianGruen** [comment]
Latest proposals: `rest:register($key)`, `rest:cancel($key)` and (as a convenience function) `rest:cancel-and-register($key)`. An annotation `%rest:register('key')` would be reasonable as well, because the functionality will usually be invoked at the very beginning of a RESTXQ function (but we’d possibly be restricted if we wanted to have a variable key).

**Comment by ChristianGruen** [comment]
The final solution: http://docs.basex.org/wiki/RESTXQ#Query_Execution

**Comment by ChristianGruen** [comment]
The Async Module is now available: #1272.
